### PR TITLE
Block Writer uses Postgres protocol

### PR DESCRIPTION
**Pull request by mrtracy**
This commit modifies the block writer example to connect to cockroach using the
postgres wire protocol.

As an additional extra, this commit also allows the block writer to connect to a
postgres database. The sql dialect for postgres is different from that supported
by cockroach db, so usage of the 'sql-dialect' flag is necessary in order to
adjust the logic being used.

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/examples-go/29)

### CODE REVIEW

**Comment by tschottdorf** [comment]
Kind of alarming that such simple operations require dialect awareness. Can't we make them use the same? What are the differences?

File: block_writer/main.go [hunk 1]
```diff
@@ -51,6 +47,146 @@ var outputInterval = flag.Duration("output-interval", 1*time.Second, "Interval o
...
+	q.Set("dbname", "postgres")
+	parsedURL.RawQuery = q.Encode()
+	parsedURL.RawPath = ""
+
+	// Open connection to the 'postgres' database.
+	pl.db, err = sql.Open("postgres", parsedURL.String())
+	if err != nil {
+		return err
+	}
+	// There is no conditional database creation in postgres; we must first
```
Comment (bdarnell) [inline block_writer/main.go]: There's no conditional creation, but there is a conditional drop, so you could do `DROP DATABASE IF EXISTS datablocks` and then an unconditional create.

File: block_writer/main.go [hunk 2]
```diff
@@ -51,6 +47,146 @@ var outputInterval = flag.Duration("output-interval", 1*time.Second, "Interval o
...
+	// Create the 'blocks' table, dropping it if it already existed.
+	if _, err := cl.db.Exec(`DROP TABLE IF EXISTS blocks`); err != nil {
+		return err
+	}
+	if _, err := cl.db.Exec(`
+		CREATE TABLE IF NOT EXISTS blocks (
+		  block_id BIGINT NOT NULL,
+		  writer_id TEXT NOT NULL,
+		  block_num BIGINT NOT NULL,
+		  raw_bytes BYTES NOT NULL,
```
Comment (bdarnell) [inline block_writer/main.go]: We now support BYTEA as an alias for BYTES, so these can be made compatible.

**Comment by mrtracy** [comment]
The three things that were incompatible:

1. Conditional database creation, not available in postgres
2. Bytea vs Bytes, I guess this is fixed
3. The insert statement in cockroach appears to require the table name; it has to be "INSERT INTO datablocks.blocks", or else I get the error "no database specified".  Conversely, the statement in postgres CANNOT use the fully qualified 'datablocks.blocks', because it assumes that is a schema.

**Comment by mrtracy** [comment]
Regarding #3: Is something messed up with my connection string?  I'm very clearly connecting with the 'dbname=datablocks' option, but it still requires me to specify a database name.

**Comment by bdarnell** [comment]
Yeah, `#3` seems like a bug on our side; we should be picking up the database from the URL.

**Comment by tschottdorf** [comment]
Ok. Looks like with a little bit of work, we could find something that works for both?

**Comment by andreimatei** [comment]
Matt, mind filing a bug for #3 with repro instructions? Sounds bad.

On Thu, Feb 11, 2016 at 7:50 PM, Tobias Schottdorf <notifications@github.com

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/examples-go/29)*

**Comment by tamird** [comment]
Replaced by #38.

File: block_writer/main.go [hunk 3]
```diff
@@ -51,6 +47,146 @@ var outputInterval = flag.Duration("output-interval", 1*time.Second, "Interval o
...
+	if _, err = cl.db.Exec("CREATE DATABASE IF NOT EXISTS datablocks"); err != nil {
+		return err
+	}
+	if err := cl.db.Close(); err != nil {
+		return err
+	}
+
+	// Open connection to the datablocks database.
+	q = parsedURL.Query()
+	q.Add("dbname", "datablocks")
```
Comment (dt) [inline block_writer/main.go]: fwiw, I think you wanted `parsedURL.Path =  "datablocks"`

Comment (petermattis) [inline block_writer/main.go]: I believe both `dbname` and `database` work as query parameters as well.

**Pull request closed without merging**